packages/web/src/common/store/cache/tracks/sagas.ts

```diff
@@ -158,6 +164,61 @@ function* editTrackAsync(action: ReturnType<typeof trackActions.editTrack>) {
     }
   }
 
+  const getFeatureEnabled = yield* getContext('getFeatureEnabled')
+  const isEditTrackRedesignEnabled = yield* call(
+    getFeatureEnabled,
+    FeatureFlags.EDIT_TRACK_REDESIGN
+  )
+  if (isEditTrackRedesignEnabled && track.stems) {
+    const currentStemUploads = yield* select(getCurrentUploads, track.track_id)
+    const existingStems = currentTrack._stems || []
+
+    // upload net new stems
+    const addedStems = track.stems.filter((stem) => {
+      return !existingStems.find((existingStem) => {
+        return (
+          existingStem.track_id === stem.metadata.track_id &&
+          existingStem.category === stem.category
+        )
+      })
+    })
+
+    const addedStemsWithFiles = addedStems.filter(
+      (stem) => 'file' in stem
+    ) as StemUploadWithFile[]
+
+    if (addedStemsWithFiles.length > 0) {
+      yield* put(
+        startStemUploads({
+          parentId: track.track_id,
+          uploads: addedStemsWithFiles,
+          batchUID: uuid()
+        })
+      )
+    }
+
+    // delete removed stems
+    const removedStems = existingStems
+      .filter((existingStem) => {
+        return !track.stems?.find(
+          (stem) =>
+            stem.metadata.track_id === existingStem.track_id &&
+            stem.category === existingStem.category
+        )
+      })
+      .filter((existingStem) => {
+        return !currentStemUploads.find(
```

Comment:
my brain is getting wonky thinking about this async stuff - are these saga functions atomic? what if an in-progress upload completes in the middle of this function?

Reply:
yeah I mostly added this for safety, but I don't think the user can even remove an in progress upload in the form 🤷